// react-stopwatch/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _utils = require('./utils');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ReactStopWatch = function (_PureComponent) {
  _inherits(ReactStopWatch, _PureComponent);

  function ReactStopWatch(props) {
    _classCallCheck(this, ReactStopWatch);

    var _this = _possibleConstructorReturn(this, (ReactStopWatch.__proto__ || Object.getPrototypeOf(ReactStopWatch)).call(this, props));

    var hours = props.hours,
        minutes = props.minutes,
        seconds = props.seconds;


    var text = (0, _utils.fillTextWithFormat)(hours, minutes, seconds);

    _this.intervalId = setInterval(function () {}, Number.MAX_VALUE);
    _this.state = {
      text: text,
      stateHours: hours,
      stateMinutes: minutes,
      stateSeconds: seconds
    };
    return _this;
  }

  _createClass(ReactStopWatch, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps() {
      this.intervalId = setInterval(function () {}, Number.MAX_VALUE);
      this.setState({
        stateHours: 0,
        stateMinutes: 0,
        stateSeconds: 0
      });
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.timer();
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.stopToCount();
    }
  }, {
    key: 'stopToCount',
    value: function stopToCount() {
      clearInterval(this.intervalId);
    }
  }, {
    key: 'counter',
    value: function counter() {
      var _this2 = this;

      var _props = this.props,
          limit = _props.limit,
          withLoop = _props.withLoop,
          onCallback = _props.onCallback;
      var _state = this.state,
          stateHours = _state.stateHours,
          stateMinutes = _state.stateMinutes,
          stateSeconds = _state.stateSeconds;


      this.setState(function (prevState) {
        return {
          stateSeconds: prevState.stateSeconds + 1
        };
      });

      if (stateSeconds >= 59) {
        this.setState(function (prevState) {
          return {
            stateSeconds: 0,
            stateMinutes: prevState.stateMinutes + 1
          };
        });

        if (stateMinutes >= 59) {
          this.setState(function (prevState) {
            return {
              stateMinutes: 0,
              stateHours: prevState.stateHours + 1
            };
          });
        }
      }

      var text = (0, _utils.fillTextWithFormat)(stateHours, stateMinutes, stateSeconds);

      this.setState({ text: text });

      if (limit === text) {
        setTimeout(function () {
          _this2.stopToCount();
          if (withLoop) {
            _this2.timer();
          }
          onCallback();
        }, 0);
      }
    }
  }, {
    key: 'timer',
    value: function timer() {
      var _this3 = this;

      var intervalId = setInterval(function () {
        _this3.counter();
      }, 1000);

      this.intervalId = intervalId;
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          children = _props2.children,
          render = _props2.render;
      var _state2 = this.state,
          text = _state2.text,
          hours = _state2.stateHours,
          minutes = _state2.stateMinutes,
          seconds = _state2.stateSeconds;


      if (children) {
        return children({
          text: text,
          hours: hours,
          minutes: minutes,
          seconds: seconds
        });
      }

      if (render) {
        return render({
          text: text,
          hours: hours,
          minutes: minutes,
          seconds: seconds
        });
      }

      console.error('Component ReactStopWatch:', 'no children or render prop are present');

      return null;
    }
  }]);

  return ReactStopWatch;
}(_react.PureComponent);

ReactStopWatch.displayName = 'ReactStopWatch';
ReactStopWatch.defaultProps = {
  withLoop: false,
  onCallback: function onCallback() {}
};
exports.default = ReactStopWatch;
